File: src/ingest/plan.ts
import fsp from 'node:fs/promises';
import type { BigIntStats, Dirent } from 'node:fs';
import path from 'node:path';
import type { FileManifestRow } from '../store/writes.js';

export type Source = 'claude' | 'codex';

export interface DiscoveredFile {
  path: string;
  source: Source;
  project: string | null;
  size: number;
  mtimeNs: bigint;
}

export interface Discovery {
  files: DiscoveredFile[];
  /** Paths whose containing dir looks unchanged; trust the manifest. */
  unchangedPaths: Set<string>;
}

export interface IngestPlan {
  toSkip: string[];
  toTail: Array<{ file: DiscoveredFile; fromOffset: number }>;
  toFullParse: DiscoveredFile[];
  toPurge: string[];
}

export interface PlanInput {
  manifest: Map<string, FileManifestRow>;
  discovery: Discovery;
  safetyWindowMs?: number;
  now?: number;
}

function manifestMtime(row: FileManifestRow): bigint {
  return typeof row.mtime_ns === 'bigint' ? row.mtime_ns : BigInt(row.mtime_ns);
}

export function planIngest(input: PlanInput): IngestPlan {
  const safetyWindowMs = input.safetyWindowMs ?? 60 * 60 * 1000;
  const now = input.now ?? Date.now();
  const safetyThresholdNs = BigInt(now - safetyWindowMs) * 1_000_000n;

  const plan: IngestPlan = { toSkip: [], toTail: [], toFullParse: [], toPurge: [] };
  const touched = new Set<string>();

  for (const d of input.discovery.files) {
    touched.add(d.path);
    const row = input.manifest.get(d.path);

    if (!row) {
      plan.toFullParse.push(d);
      continue;
    }

    const rowMtime = manifestMtime(row);

    if (d.size < row.last_offset) {
      plan.toFullParse.push(d);
      continue;
    }

    if (d.size === row.size && d.mtimeNs === rowMtime) {
      plan.toSkip.push(d.path);
      continue;
    }

    if (d.size > row.last_offset) {
      plan.toTail.push({ file: d, fromOffset: row.last_offset });
    } else {
      plan.toSkip.push(d.path);
    }
  }

  for (const p of input.discovery.unchangedPaths) {
    if (touched.has(p)) continue;
    const row = input.manifest.get(p);
    if (!row) continue;
    const rowMtime = manifestMtime(row);
    if (rowMtime >= safetyThresholdNs) {
      // Spec: recent files are always re-stat'd even if dir looks unchanged.
      // Caller guarantees this case doesn't arrive via unchangedPaths — if it
      // does anyway, demote to a full-parse-safe skip: keep the data as-is.
      plan.toSkip.push(p);
    } else {
      plan.toSkip.push(p);
    }
    touched.add(p);
  }

  for (const [p] of input.manifest) {
    if (!touched.has(p)) plan.toPurge.push(p);
  }

  return plan;
}

// --- discovery helpers ---

function decodeClaudeSlug(folderName: string): string | null {
  if (!folderName.startsWith('-')) return null;
  return '/' + folderName.slice(1).replace(/-/g, '/');
}

async function statDir(dir: string): Promise<BigIntStats | null> {
  try {
    return await fsp.stat(dir, { bigint: true });
  } catch {
    return null;
  }
}

async function listDir(dir: string): Promise<Dirent[]> {
  try {
    return await fsp.readdir(dir, { withFileTypes: true });
  } catch {
    return [];
  }
}

// Index the manifest by parent directory once per run so dir-level lookups are O(1).
export interface ManifestDirIndex {
  byParent: Map<string, { paths: string[]; maxMtime: bigint }>;
}

export function indexManifestByParent(
  manifest: Map<string, FileManifestRow>,
): ManifestDirIndex {
  const byParent = new Map<string, { paths: string[]; maxMtime: bigint }>();
  for (const [p, row] of manifest) {
    const parent = path.dirname(p);
    let entry = byParent.get(parent);
    if (!entry) {
      entry = { paths: [], maxMtime: 0n };
      byParent.set(parent, entry);
    }
    entry.paths.push(p);
    const m = manifestMtime(row);
    if (m > entry.maxMtime) entry.maxMtime = m;
  }
  return { byParent };
}

interface DiscoverOptions {
  manifest: Map<string, FileManifestRow>;
  manifestIndex?: ManifestDirIndex;
  safetyWindowMs?: number;
  now?: number;
}

function isRecentMtime(mtimeNs: bigint, safetyThresholdNs: bigint): boolean {
  return mtimeNs >= safetyThresholdNs;
}

async function walkAllFiles(
  dir: string,
  accept: (absPath: string, size: number, mtimeNs: bigint) => void | Promise<void>,
): Promise<void> {
  for (const e of await listDir(dir)) {
    const abs = path.join(dir, e.name);
    if (e.isDirectory()) {
      await walkAllFiles(abs, accept);
    } else if (e.isFile() && e.name.endsWith('.jsonl')) {
      try {
        const st = await fsp.stat(abs, { bigint: true });
        await accept(abs, Number(st.size), st.mtimeNs);
      } catch {
        // race; skip
      }
    }
  }
}

/**
 * Walks a directory with dir-mtime short-circuit. If the directory's mtime is
 * older than any manifest mtime it contains (and no file under it is "recent"),
 * skip statting individual files and mark the manifest paths under it as
 * unchanged.
 */
async function walkWithShortCircuit(
  dir: string,
  source: Source,
  project: string | null,
  opts: DiscoverOptions,
  safetyThresholdNs: bigint,
  out: Discovery,
): Promise<void> {
  const st = await statDir(dir);
  if (!st) return;
  const dirMtime = st.mtimeNs;

  const index = opts.manifestIndex;
  const entry = index?.byParent.get(dir);
  // If this dir has manifest entries and the dir mtime hasn't advanced past
  // them, skip statting files inside.
  if (entry && dirMtime <= entry.maxMtime) {
    let hasRecent = false;
    for (const p of entry.paths) {
      const row = opts.manifest.get(p);
      if (!row) continue;
      if (isRecentMtime(manifestMtime(row), safetyThresholdNs)) {
        hasRecent = true;
        break;
      }
    }
    if (!hasRecent) {
      for (const p of entry.paths) out.unchangedPaths.add(p);
      return;
    }
    // Fall through: there's a recent file; do the full walk.
  }

  // Full walk: stat every file.
  for (const e of await listDir(dir)) {
    const abs = path.join(dir, e.name);
    if (e.isDirectory()) {
      await walkWithShortCircuit(abs, source, project, opts, safetyThresholdNs, out);
    } else if (e.isFile() && e.name.endsWith('.jsonl')) {
      try {
        const fs_ = await fsp.stat(abs, { bigint: true });
        out.files.push({
          path: abs,
          source,
          project,
          size: Number(fs_.size),
          mtimeNs: fs_.mtimeNs,
        });
      } catch {
        // race; skip
      }
    }
  }
}

export async function discoverClaude(
  roots: string[],
  opts: DiscoverOptions,
): Promise<Discovery> {
  const now = opts.now ?? Date.now();
  const safetyWindowMs = opts.safetyWindowMs ?? 60 * 60 * 1000;
  const safetyThresholdNs = BigInt(now - safetyWindowMs) * 1_000_000n;
  const resolvedOpts: DiscoverOptions = {
    ...opts,
    manifestIndex: opts.manifestIndex ?? indexManifestByParent(opts.manifest),
  };

  const discovery: Discovery = { files: [], unchangedPaths: new Set() };
  for (const root of roots) {
    for (const e of await listDir(root)) {
      if (!e.isDirectory()) continue;
      const projectDir = path.join(root, e.name);
      const project = decodeClaudeSlug(e.name);
      await walkWithShortCircuit(
        projectDir,
        'claude',
        project,
        resolvedOpts,
        safetyThresholdNs,
        discovery,
      );
    }
  }
  return discovery;
}

export async function discoverCodex(
  roots: string[],
  opts: DiscoverOptions,
): Promise<Discovery> {
  const now = opts.now ?? Date.now();
  const safetyWindowMs = opts.safetyWindowMs ?? 60 * 60 * 1000;
  const safetyThresholdNs = BigInt(now - safetyWindowMs) * 1_000_000n;
  const resolvedOpts: DiscoverOptions = {
    ...opts,
    manifestIndex: opts.manifestIndex ?? indexManifestByParent(opts.manifest),
  };

  const discovery: Discovery = { files: [], unchangedPaths: new Set() };
  for (const root of roots) {
    await walkWithShortCircuit(
      root,
      'codex',
      null,
      resolvedOpts,
      safetyThresholdNs,
      discovery,
    );
  }
  return discovery;
}

export async function discoverAll(input: {
  claudeRoots: string[];
  codexRoots: string[];
  manifest: Map<string, FileManifestRow>;
  safetyWindowMs?: number;
  now?: number;
}): Promise<Discovery> {
  const [c, x] = await Promise.all([
    discoverClaude(input.claudeRoots, {
      manifest: input.manifest,
      safetyWindowMs: input.safetyWindowMs,
      now: input.now,
    }),
    discoverCodex(input.codexRoots, {
      manifest: input.manifest,
      safetyWindowMs: input.safetyWindowMs,
      now: input.now,
    }),
  ]);
  const merged: Discovery = {
    files: [...c.files, ...x.files],
    unchangedPaths: new Set<string>([...c.unchangedPaths, ...x.unchangedPaths]),
  };
  return merged;
}

// Legacy discovery path for --no-cache mode: no short-circuit, no manifest.
export async function discoverClaudeLegacy(roots: string[]): Promise<DiscoveredFile[]> {
  const out: DiscoveredFile[] = [];
  for (const root of roots) {
    for (const e of await listDir(root)) {
      if (!e.isDirectory()) continue;
      const projectDir = path.join(root, e.name);
      const project = decodeClaudeSlug(e.name);
      await walkAllFiles(projectDir, (abs, size, mtimeNs) => {
        out.push({ path: abs, source: 'claude', project, size, mtimeNs });
      });
    }
  }
  return out;
}

export async function discoverCodexLegacy(roots: string[]): Promise<DiscoveredFile[]> {
  const out: DiscoveredFile[] = [];
  for (const root of roots) {
    await walkAllFiles(root, (abs, size, mtimeNs) => {
      out.push({ path: abs, source: 'codex', project: null, size, mtimeNs });
    });
  }
  return out;
}

File: src/ingest/plan.rs
use crate::discovery::{DiscoveredFile, Discovery};
use crate::store::writes::FileManifestRow;
use crate::types::Source;
use std::collections::{HashMap, HashSet};
use std::path::PathBuf;

#[derive(Debug, Default)]
pub struct IngestPlan {
    pub to_skip: Vec<PathBuf>,
    pub to_tail: Vec<TailItem>,
    pub to_full_parse: Vec<DiscoveredFile>,
    pub to_purge: Vec<PathBuf>,
}

#[derive(Debug)]
pub struct TailItem {
    pub file: DiscoveredFile,
    pub from_offset: u64,
}

pub struct PlanInput<'a> {
    pub manifest: &'a HashMap<PathBuf, FileManifestRow>,
    pub discovery: &'a Discovery,
    pub safety_window_ms: i64,
    pub now_ms: i64,
}

pub fn plan_ingest(input: PlanInput<'_>) -> IngestPlan {
    let safety_threshold_ns = (input.now_ms - input.safety_window_ms).saturating_mul(1_000_000);

    let mut plan = IngestPlan::default();
    let mut touched: HashSet<PathBuf> = HashSet::new();

    for d in &input.discovery.files {
        touched.insert(d.path.clone());
        let Some(row) = input.manifest.get(&d.path) else {
            plan.to_full_parse.push(d.clone());
            continue;
        };

        // Shrunk/rotated file → full parse
        if d.size < row.last_offset {
            plan.to_full_parse.push(d.clone());
            continue;
        }

        // Recently modified file → full parse (don't risk partial tail)
        if d.mtime_ns >= safety_threshold_ns {
            // Even if it looks unchanged, re-parse inside safety window to be safe.
            // But if truly identical (same size + mtime) we can still skip.
            if d.size == row.size && d.mtime_ns == row.mtime_ns {
                plan.to_skip.push(d.path.clone());
            } else {
                plan.to_full_parse.push(d.clone());
            }
            continue;
        }

        // Unchanged
        if d.size == row.size && d.mtime_ns == row.mtime_ns {
            plan.to_skip.push(d.path.clone());
            continue;
        }

        if d.source == Source::Cursor {
            plan.to_full_parse.push(d.clone());
            continue;
        }

        // Grown, safe to tail
        if d.size > row.last_offset {
            plan.to_tail.push(TailItem {
                file: d.clone(),
                from_offset: row.last_offset,
            });
        } else {
            plan.to_skip.push(d.path.clone());
        }
    }

    // Files in unchanged directories → skip (manifest is trusted)
    for p in &input.discovery.unchanged_paths {
        if touched.contains(p) {
            continue;
        }
        if input.manifest.contains_key(p) {
            plan.to_skip.push(p.clone());
            touched.insert(p.clone());
        }
    }

    // Anything in manifest but not discovered → purge
    for p in input.manifest.keys() {
        if !touched.contains(p) {
            plan.to_purge.push(p.clone());
        }
    }

    plan
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::Source;
    use std::path::PathBuf;

    fn mk_file(path: &str, size: u64, mtime: i64) -> DiscoveredFile {
        DiscoveredFile {
            path: PathBuf::from(path),
            source: Source::Claude,
            project: None,
            size,
            mtime_ns: mtime,
        }
    }

    fn mk_row(path: &str, size: u64, mtime: i64, last_offset: u64) -> FileManifestRow {
        FileManifestRow {
            path: PathBuf::from(path),
            source: Source::Claude,
            project: None,
            size,
            mtime_ns: mtime,
            last_offset,
            n_events: 0,
            session_id: None,
            model: None,
        }
    }

    fn far_past_ns() -> i64 {
        // 2020-01-01 in nanoseconds — well outside any test's safety window.
        1_577_836_800_000_000_000
    }

    fn base_now_ms() -> i64 {
        // 2026-04-22 roughly
        1_777_000_000_000
    }

    #[test]
    fn new_file_goes_to_full_parse() {
        let files = vec![mk_file("/x.jsonl", 100, far_past_ns())];
        let discovery = Discovery {
            files,
            unchanged_paths: Default::default(),
        };
        let manifest = HashMap::new();
        let plan = plan_ingest(PlanInput {
            manifest: &manifest,
            discovery: &discovery,
            safety_window_ms: 3_600_000,
            now_ms: base_now_ms(),
        });
        assert_eq!(plan.to_full_parse.len(), 1);
        assert!(plan.to_tail.is_empty());
        assert!(plan.to_skip.is_empty());
    }

    #[test]
    fn unchanged_file_goes_to_skip() {
        let mt = far_past_ns();
        let files = vec![mk_file("/x.jsonl", 500, mt)];
        let discovery = Discovery {
            files,
            unchanged_paths: Default::default(),
        };
        let mut manifest = HashMap::new();
        manifest.insert(PathBuf::from("/x.jsonl"), mk_row("/x.jsonl", 500, mt, 500));
        let plan = plan_ingest(PlanInput {
            manifest: &manifest,
            discovery: &discovery,
            safety_window_ms: 3_600_000,
            now_ms: base_now_ms(),
        });
        assert_eq!(plan.to_skip.len(), 1);
        assert!(plan.to_full_parse.is_empty());
    }

    #[test]
    fn grown_file_goes_to_tail() {
        let mt = far_past_ns();
        let files = vec![mk_file("/x.jsonl", 800, mt + 1)];
        let discovery = Discovery {
            files,
            unchanged_paths: Default::default(),
        };
        let mut manifest = HashMap::new();
        manifest.insert(PathBuf::from("/x.jsonl"), mk_row("/x.jsonl", 500, mt, 500));
        let plan = plan_ingest(PlanInput {
            manifest: &manifest,
            discovery: &discovery,
            safety_window_ms: 3_600_000,
            now_ms: base_now_ms(),
        });
        assert_eq!(plan.to_tail.len(), 1);
        assert_eq!(plan.to_tail[0].from_offset, 500);
    }

    #[test]
    fn shrunk_file_goes_to_full_parse() {
        let mt = far_past_ns();
        let files = vec![mk_file("/x.jsonl", 100, mt + 1)];
        let discovery = Discovery {
            files,
            unchanged_paths: Default::default(),
        };
        let mut manifest = HashMap::new();
        manifest.insert(PathBuf::from("/x.jsonl"), mk_row("/x.jsonl", 500, mt, 500));
        let plan = plan_ingest(PlanInput {
            manifest: &manifest,
            discovery: &discovery,
            safety_window_ms: 3_600_000,
            now_ms: base_now_ms(),
        });
        assert_eq!(plan.to_full_parse.len(), 1);
    }

    #[test]
    fn disappeared_file_goes_to_purge() {
        let discovery = Discovery::default();
        let mut manifest = HashMap::new();
        manifest.insert(
            PathBuf::from("/gone.jsonl"),
            mk_row("/gone.jsonl", 100, 1, 100),
        );
        let plan = plan_ingest(PlanInput {
            manifest: &manifest,
            discovery: &discovery,
            safety_window_ms: 3_600_000,
            now_ms: base_now_ms(),
        });
        assert_eq!(plan.to_purge, vec![PathBuf::from("/gone.jsonl")]);
    }

    #[test]
    fn recent_file_goes_to_full_parse_even_if_grown() {
        // mtime within safety window
        let now_ms = base_now_ms();
        let recent_ns = (now_ms - 1_000).saturating_mul(1_000_000);
        let files = vec![mk_file("/x.jsonl", 800, recent_ns)];
        let discovery = Discovery {
            files,
            unchanged_paths: Default::default(),
        };
        let mut manifest = HashMap::new();
        manifest.insert(
            PathBuf::from("/x.jsonl"),
            mk_row("/x.jsonl", 500, recent_ns - 1000, 500),
        );
        let plan = plan_ingest(PlanInput {
            manifest: &manifest,
            discovery: &discovery,
            safety_window_ms: 3_600_000,
            now_ms,
        });
        assert_eq!(plan.to_full_parse.len(), 1);
        assert!(plan.to_tail.is_empty());
    }

    #[test]
    fn unchanged_directory_skips_file_without_stat() {
        let discovery = Discovery {
            files: vec![],
            unchanged_paths: {
                let mut s = HashSet::new();
                s.insert(PathBuf::from("/dir/x.jsonl"));
                s
            },
        };
        let mut manifest = HashMap::new();
        manifest.insert(
            PathBuf::from("/dir/x.jsonl"),
            mk_row("/dir/x.jsonl", 100, far_past_ns(), 100),
        );
        let plan = plan_ingest(PlanInput {
            manifest: &manifest,
            discovery: &discovery,
            safety_window_ms: 3_600_000,
            now_ms: base_now_ms(),
        });
        assert_eq!(plan.to_skip, vec![PathBuf::from("/dir/x.jsonl")]);
        assert!(plan.to_purge.is_empty());
    }

    #[test]
    fn changed_cursor_file_goes_to_full_parse_not_tail() {
        let mt = far_past_ns();
        let discovery = Discovery {
            files: vec![DiscoveredFile {
                path: PathBuf::from("/cursor/usage.csv"),
                source: Source::Cursor,
                project: None,
                size: 800,
                mtime_ns: mt + 1,
            }],
            unchanged_paths: Default::default(),
        };
        let mut manifest = HashMap::new();
        manifest.insert(
            PathBuf::from("/cursor/usage.csv"),
            FileManifestRow {
                path: PathBuf::from("/cursor/usage.csv"),
                source: Source::Cursor,
                project: None,
                size: 500,
                mtime_ns: mt,
                last_offset: 500,
                n_events: 0,
                session_id: None,
                model: None,
            },
        );
        let plan = plan_ingest(PlanInput {
            manifest: &manifest,
            discovery: &discovery,
            safety_window_ms: 3_600_000,
            now_ms: base_now_ms(),
        });
        assert_eq!(plan.to_full_parse.len(), 1);
        assert!(plan.to_tail.is_empty());
    }
}
